discovery/sync_manager.go

```diff
@@ -0,0 +1,619 @@
+	ChanSeries ChannelGraphTimeSeries
+
+	// NumActiveSyncers is the number of peers for which we should have
+	// active syncers with. After reaching NumActiveSyncers, any future
+	// gossip syncers will be passive.
+	NumActiveSyncers int
+
+	// RotateTicker is a ticker responsible for notifying the syncManager
+	// when it should rotate its active syncers. Any active syncers with a
+	// chansSynced state will be exchanged for a passive syncer in order to
+	// ensure we don't keep syncing with the same peers.
+	RotateTicker ticker.Ticker
+
+	// FullSyncTicker is a ticker responsible for notifying the syncManager
+	// when it should attempt a full sync with a gossip sync peer.
+	FullSyncTicker ticker.Ticker
+
+	// ActiveSyncerTimeoutTicker is a ticker responsible for notifying the
+	// syncManager when it should attempt to start the next pending
+	// activeSyncer due to the current one not completing its state machine
+	// within the timeout.
+	ActiveSyncerTimeoutTicker ticker.Ticker
+}
+
+// activeSyncer is an internal wrapper type over a gossipSyncer used within the
+// syncManager to determine whether a sync transition to an activeSync type
+// should be performed before attempting to start the syncer.
+type activeSyncer struct {
+	*gossipSyncer
+	transition bool
+	errChan    chan error
+}
+
+// syncManager is a subsystem of the gossiper that manages the gossip syncers
+// for peers currently connected. When a new peer is connected, the manager will
+// create its accompanying gossip syncer and determine whether it should have an
+// activeSync or passiveSync sync type based on how many other gossip syncers
+// are currently active. Any activeSync gossip syncers are started in a
+// round-robin manner to ensure we're not syncing with multiple peers at the
+// same time.
+type syncManager struct {
+	start sync.Once
+	stop  sync.Once
+
+	cfg syncManagerCfg
+
+	// syncersMtx guards the access to the variables activeSyncers and
+	// syncers below.
+	syncersMtx sync.RWMutex
+
+	// activeSyncers is the set of all syncers for which we are currently
+	// receiving graph updates from. The number of possible active syncers
+	// is bounded by NumActiveSyncers.
+	activeSyncers map[routing.Vertex]*gossipSyncer
+
+	// syncers is the set of all syncers for which we are not currently
+	// receiving graph updates from.
+	syncers map[routing.Vertex]*gossipSyncer
+
+	// newActiveSyncers is a channel through which we'll send any new active
+	// syncers that should be started in a round-robin manner.
+	newActiveSyncers chan *activeSyncer
+
+	// staleActiveSyncers is a channel through which we'll send any stale
+	// active syncers that should be removed from the round-robin.
+	staleActiveSyncers chan routing.Vertex
+
+	// pendingActiveSyncers is the list of active syncers which are pending
+	// to be started. Syncers will be added to this list through the
+	// newActiveSyncers and staleActiveSyncers channels.
+	pendingActiveSyncers []*activeSyncer
```

Comment:
If this is made into a map, then slice tricks aren't required below when removing stale peers from this structure.

Reply:
Fixed.